### PR TITLE
US136043: Properties for Google Drive can be edited and saved

### PR DESCRIPTION
**Pull request by coffeexcode**
## Relevant Rally
- [US136043](https://rally1.rallydev.com/#/?detail=/userstory/626774144995&fdp=true): Properties for Google Drive can be edited and saved

## Description
This PR updates the weblink editor to support different types of weblinks, specifically for google drive and onedrive links (referred to as `cloud drives`. This PR also decouples the display options from the link and should still have the same error handling as before.

## Testing
- [x] Another developer will test this

You will need the following
- LMS: Master >= Feb 16
- Lessons (likely running locally): 22.2.1
- Dev tunnel enabled for drive pickers

## OneDrive
image
image

## Google Drive
image
image

## Plain Weblinks
image
image

### CODE REVIEW

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 1]
```diff
@@ -71,21 +192,25 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
 				.href="${this.activityUsageHref}"
 				.token="${this.token}"
 				?skeleton="${this.skeleton}"
-				expanded="true"
+				?expanded="${true}"
```
Comment (brissons) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:194]: nit: this can be simplified to just `expanded`

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 2]
```diff
@@ -158,11 +290,127 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+				composed: true,
+				detail: {
+					title: event.m_title,
+					url: event.m_url
+				}
+			}));
+		});
+	}
+
+	_openPopout() {
```
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:395]: Oh right. I will make sure it does what LTI does (when you click the button - the preview goes away forever, which I agree is weird)

File: components/d2l-google-drive-viewer/d2l-google-driver-viewer.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { EntityMixinLit } from 'siren-sdk/src/mixin/entity-mixin-lit';
+
+const ifrauImportSrc = `https://s.brightspace.com/lib/ifrau/0.28.0/ifrau/host.js?cacheBuster=${Date.now()}`;
```
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:4]: I can do that.

File: components/d2l-activity-editor/d2l-activity-content-editor/d2l-activity-content-editor-detail.js [hunk 4]
```diff
@@ -74,11 +74,16 @@ class ContentEditorDetail extends MobxLitElement {
 		}
 
 		if (entityType === CONTENT_TYPES.weblink) {
+			const queryParams = new Proxy(new URLSearchParams(window.location.search), {
```
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/d2l-activity-content-editor-detail.js:77]: Right. I think I forgot to handle this case. I had planned on basically using the query param (because I think route parsing is less nice) to help load the initial skeleton, the change the type based on the serialized response (which means you wouldn't be able to use the page with the pickers when the flag is off).

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 5]
```diff
@@ -60,6 +103,84 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+				<d2l-activity-content-web-link-url-preview
+					.entity=${webLinkEntity}
+					?skeleton="${this.skeleton}"
+				>
+				</d2l-activity-content-web-link-url-preview>
+			`;
+		} else {
+			let name = '';
+
+			if (webLinkEntity) {
```
[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:131]: oh wow i definitely thought there was a ! when there wasn't. ignore me!

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 6]
```diff
@@ -158,11 +290,127 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+	_openGoogleDriveLinkDialog() {
+		const opener = D2L.LP.Web.UI.ObjectRepository.TryGet('D2L.ActivityEditor.AddGoogleDriveLinkDialogOpener');
+		if (!opener) {
+			return;
+		}
+
+		const event = opener();
+
+		event.AddListener(event => {
+			this.dispatchEvent(new CustomEvent('googledrive-link-dialog-opened', {
```
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:350]: Hmm. I'll see if I can play around with the event structure and clean them up. I found the `save` events to be really confusing when there were multiple places that all handled the same event and was trying to make it easier to trace down actions that triggered the events.

File: components/d2l-activity-editor/d2l-activity-content-editor/d2l-activity-content-editor-detail.js [hunk 7]
```diff
@@ -74,11 +75,21 @@ class ContentEditorDetail extends MobxLitElement {
 		}
 
 		if (entityType === CONTENT_TYPES.weblink) {
+			let type;
+			if (window.location.href.includes('googledrive')) {
+				type = WebLinkTypes.GoogleDrive;
+			} else if (window.location.href.includes('onedrive')) {
+				type = WebLinkTypes.OneDrive;
+			} else {
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/d2l-activity-content-editor-detail.js:83]: code style suggestion, if you wanted you could just remove this `else` entire if you initialize your `type = WebLinkTypes.Default` at the beginning

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-onedrive-link-jump-icon.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+							<g fill-rule="nonzero">
+								<path fill="#006FBF" d="M250.047 94.72l-13.67-16A7.803 7.803 0 0 0 230.49 76c-.958 0-1.908.183-2.797.54a7.992 7.992 0 0 0-5.026 7.46v10h-52.801a6.001 6.001 0 0 0 0 12h52.801v10a8.005 8.005 0 0 0 5.085 7.48 7.34 7.34 0 0 0 2.738.52 7.694 7.694 0 0 0 5.886-2.74l13.69-15.98a8.144 8.144 0 0 0-.02-10.56zm-2.934 7.92l-13.689 16a3.83 3.83 0 0 1-2.933 1.36 3.621 3.621 0 0 1-1.37-.26 3.998 3.998 0 0 1-2.542-3.74v-12a1.984 1.984 0 0 0-1.955-2h-54.757a2 2 0 0 1 0-4h54.757a1.984 1.984 0 0 0 1.956-2V84a3.998 3.998 0 0 1 2.542-3.74c.435-.175.9-.263 1.369-.26a3.86 3.86 0 0 1 2.933 1.36l13.69 16a4.038 4.038 0 0 1 0 5.28z"/>
+								<path fill="#FFF" d="M247.11 102.64l-13.69 16a3.83 3.83 0 0 1-2.933 1.36 3.621 3.621 0 0 1-1.369-.26 3.998 3.998 0 0 1-2.542-3.74v-12a1.984 1.984 0 0 0-1.956-2h-54.757a2 2 0 0 1 0-4h54.757a1.984 1.984 0 0 0 1.956-2V84a3.998 3.998 0 0 1 2.542-3.74c.435-.175.9-.263 1.369-.26a3.86 3.86 0 0 1 2.933 1.36l13.69 16a4.038 4.038 0 0 1 0 5.28z"/>
+							</g>
+						</g>
+					</svg>
+				</div>
+				<slot></slot>
+			</div>
+
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-onedrive-link-jump-icon.js:46]: nit: some unnecessary whitespace here

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 9]
```diff
@@ -71,21 +233,23 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
 				.href="${this.activityUsageHref}"
 				.token="${this.token}"
 				?skeleton="${this.skeleton}"
-				expanded="true"
+				expanded
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:236]: is this supposed to have the `?` since it is a boolean?

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 10]
```diff
@@ -158,10 +329,106 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+		}
+
+		this.linkError = null;
+
+		const webLinkEntity = webLinkStore.getContentWebLinkActivity(this.href);
+		if (!webLinkEntity) {
+			return;
+		}
+
+		if (this.type !== WebLinkTypes.Default) {
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:344]: nit: i would just `||` this check with the above one

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+					?disabled="${!this.canEmbed}"
+					@change="${this._saveLinkOptions}"
+					aria-invalid="${!!this._linkOptionsError}">
+					${this.localize('content.embedOnPage')}
+			</label>
+			<label class="d2l-input-radio-label d2l-skeletize">
+				<input
+					id="open-new-tab"
+					type="radio"
+					name="link-display-group${!this.canEmbed ? '-1' : ''}"
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js:108]: this is a super nit, but imo this would read nicer if you flipped the ternary, then you don't have to think in negatives:
```
{this.canEmbed ? '' : '-1'}
```

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js (hunk 9 above)
Comment (ianlaidlaw) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:236]: i think you're right

File: components/d2l-google-drive-viewer/d2l-google-driver-viewer.js [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		if (!fileId) {
+			return;
+		}
+
+		const root = this.shadowRoot;
+
+		if (root.hasChildNodes()) {
+			// The FRA injects an iframe, when we change the link
+			// we need to remove the old iframe
+			root.innerHTML = '';
```
Comment (kristian-torres) [inline components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:95]: is this adequate in removing the old iframe? we don't need to `.removeChild()` or something? I'm not sure.

Comment (coffeexcode) [reply components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:95]: Yeah, works like a charm. I originally tried the .removeChild() route but just nuking the interior is less LOC than finding and removing the child, so I went with this approach

**Comment by coffeexcode** [comment]
> just noticed with the screenshots: image ...the disabled option is selected. I think the default value is "Embed on the page", even though there isn't support for it. if we can't embed, maybe we should have the "open in a new tab" selected by default?

Apparently I missed your comment, sorry Scott. Basically I was following what LTI does. When you create the weblink and set it to be embedded, that is how you get the to the state where the disabled option is selected. LTI basically does the same thing (for non-embeddable LTI links) where the user just won't be able to save any new details about the link until they change that option. I agree though, I think there should be a ticket to capture some additional changes for the creation / updating of weblinks.

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 13]
```diff
@@ -58,8 +110,118 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+						<d2l-google-drive-viewer
+							.token=${this.token}
+							.link=${link}
+						></d2l-google-drive-viewer>
+					</div>
+				`;
+			}
+
+			cloudDriveBanner = html`
+				<d2l-activity-content-external-activity-container
```
[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:181]: the same thing happens with oneDrive but persists in the mobile view

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 14]
```diff
@@ -58,8 +110,118 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+					}
+				});
+
+				this.dispatchEvent(event);
+			}
+
+			if (this.type === WebLinkTypes.OneDrive) {
+				defaultLinkPreview = html`
+					<div class="d2l-link-preview d2l-onedrive-link d2l-skeletize">
+						<d2l-activity-content-onedrive-link-jump-icon>
```
Comment (brissons) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:162]: the "Open OneDrive file in a new browser tab" text is not tab-able

File: components/d2l-google-drive-viewer/d2l-google-driver-viewer.js [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+	disconnectedCallback() {
+		this._disconnectHost();
+		window.removeEventListener('weblink-picker-url-updated', () => {});
+
+		super.disconnectedCallback();
+	}
+
+	render() {
+		return html``;
```
[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:50]: I know weblink uses working copy, so not sure if this can be avoided.

**Comment by brissons** [comment]
Dev. tested and lgtm! Just some final comments:

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 16]
```diff
@@ -37,19 +78,30 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+			if (e.detail.href) {
+				this.activityUsageHref = e.detail.href;
+			}
+		});
+		window.addEventListener('weblink-picker-url-updated', (e) => this._handlePickerClose(e));
 	}
 
 	disconnectedCallback() {
-		window.removeEventListener('d2l-refetch-complete', this.onRefetchComplete);
+		window.removeEventListener('d2l-refetch-complete', () => {});
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:103]: is this doing anything? it looks like you created an event listener with an anonymous function, i'm not sure if this will actually remove it. would it be worth it to make those functions named

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 17]
```diff
@@ -158,10 +329,83 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
+
+	_openLinkPickerDialog() {
+		const webLinkEntity = webLinkStore.getContentWebLinkActivity(this.href);
+
+		if (!webLinkEntity) {
+			return;
+		}
+
+		if (webLinkEntity.type === WebLinkTypes.GoogleDrive) {
+			return this._openLinkPicker('D2L.ActivityEditor.AddGoogleDriveLinkDialogOpener');
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:384]: just curious, is there a reason that these need to `return` ?

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 18]
```diff
@@ -58,8 +110,118 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
 		if (webLinkEntity) {
 			this.skeleton = false;
+			this.type = webLinkEntity.type;
 		}
 
+		let defaultLinkEditor = html``;
+		let defaultLinkPreview = html``;
+		let cloudDriveBanner = html``;
+
+		if (this.type === WebLinkTypes.Default) {
```
[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:120]: (I will respond to this with an answer soon!)

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+	constructor() {
+		super();
+		this._debounceJobs = {};
+		this.saveOrder = 2000;
+		this.showLinkOptions = true;
+	}
+
+	render() {
+		const isExternalResource = this.entity
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js:62]: suggestion: this could be shortened if you prefer:
```
const isExternalResource = !!( this.entity && this.entity.isExternalResource );
```

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+					?disabled="${!this.canEmbed}"
+					@change="${this._saveLinkOptions}"
+					aria-invalid="${!!this._linkOptionsError}">
+					${this.localize('content.embedOnPage')}
+			</label>
+			<label class="d2l-input-radio-label d2l-skeletize">
+				<input
+					id="open-new-tab"
+					type="radio"
+					name="link-display-group${this.canEmbed ? '' : '-1'}"
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js:108]: could you explain the scenario why the name should be `link-display-group-1` ? what is that doing?

File: components/d2l-google-drive-viewer/d2l-google-driver-viewer.js [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+			const ifrauImport = document.createElement('script');
+			ifrauImport.src = ifrauImportSrc;
+			ifrauImport.onload = () => this._onEntityChanged(this.link);
+			document.head.appendChild(ifrauImport);
+		}
+	}
+
+	disconnectedCallback() {
+		this._disconnectHost();
+		window.removeEventListener('weblink-picker-url-updated', () => {});
```
Comment (ianlaidlaw) [inline components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:44]: same question as above, does this line actually do anything?

File: components/d2l-google-drive-viewer/d2l-google-driver-viewer.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-google-drive-viewer/d2l-google-driver-viewer.js:50]: I remember trying to fix this, but I wasn't able to find a solution, the component page is re-rendering with working copy... I can try to take a peek again, but not sure there is a solution (without being a big lift)

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js (hunk 16 above)
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:103]: Hmm, yeah it looks like that is the case. I will switch the functions to named functions and update the removal.

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js (hunk 20 above)
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-display-options.js:108]: Basically there needs to be a scenario to update the embed for web links. When you create a weblink, you can select one drive (which cannot embed). This means that one of the radio buttons has to be disabled. When you disable the radio button, there was an issue with screen readers not reading out the element or being tab-able (something to do with an option being disabled). By changing the name like this, it makes the radio buttons separate groups and fixes the screen reader error.

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js (hunk 17 above)
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:384]: Not that I can remember. Removed.

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:181]: So in this case, what should we do with the buttons? Just keep shrinking the title makes the most sense to me.

[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:181]: makes sense to me too!

[the remark this answers is not in the document]
Comment (brissons) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:181]: we could double check with Diana just to make sure

File: components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js [hunk 22]
```diff
@@ -37,19 +78,26 @@ class ContentWebLinkDetail extends SkeletonMixin(ErrorHandlingMixin(LocalizeActi
...
 
-		window.addEventListener('d2l-refetch-complete', this.onRefetchComplete);
+		window.addEventListener('d2l-refetch-complete', this._handleRefetchComplete);
+		window.addEventListener('weblink-picker-url-updated', this._handlePickerClose);
 	}
 
 	disconnectedCallback() {
-		window.removeEventListener('d2l-refetch-complete', this.onRefetchComplete);
+		window.removeEventListener('d2l-refetch-complete', this._handleRefetchComplete);
+		window.removeEventListener('weblink-picker-url-updated', this._handleP);
```
Comment (ianlaidlaw) [inline components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js:100]: think you accidentally some letters

Comment (coffeexcode) [reply components/d2l-activity-editor/d2l-activity-content-editor/web-link/d2l-activity-content-web-link-detail.js]: Whoops, should be

**Review by ianlaidlaw — APPROVED** [review]
any comments i had look to be addressed!
